File: packages/catalog-schemas/src/components/lift-editor.ts
import {z} from 'zod';
import {constructionsFieldSourceSchema} from './constructions-editor';
import {formFieldBaseSchema} from './form-card';
import {type CatalogComponentDefinition} from '../types';

/**
 * Источник значения поля — тот же контракт, что `generalSources` теплотеха:
 * словарь словоформ у рендереров общий, поэтому видам источника расходиться
 * нельзя — схема одна на оба компонента, а не копия.
 */
export const liftEditorFieldSourceSchema = constructionsFieldSourceSchema;

// Источники по именам полей одного экрана (здания или лифта). Параллельный
// блок, а не обёртка над значениями: payload submit'а не меняется.
export const liftEditorSectionSourcesSchema = z.record(
  z.string(),
  liftEditorFieldSourceSchema,
);

// Режим лифтов методики: перечень вкладок «Лифт 1…N» с add/remove либо одна
// вкладка лифтовой группы (число лифтов в ней — обычное поле здания).
export const liftsModeSchema = z.enum(['per-lift', 'group']);

// Экран, на котором живёт поле-источник: значения лифта (по умолчанию) или
// значения здания. Тип здания на «Здании» управляет полями лифта.
export const liftEditorFieldScopeSchema = z.enum(['building', 'lift']);

/**
 * Поле редактора — базовое поле FormCard плюс аддитивные ключи. Новой
 * параллельной модели поля в каталоге не заводим (Решение 5 design.md).
 */
export const liftEditorFieldSchema = formFieldBaseSchema
  .extend({
    // `combo` — свободный ввод с подсказками ряда (datalist), значение вне
    // ряда допустимо: реальные каталоги лифтов выходят за ряды ГОСТ.
    type: z.enum(['text', 'number', 'select', 'boolean', 'lookup', 'combo']),
    // Поле с осмысленным значением по умолчанию — под экспандер.
    advanced: z.boolean().optional(),
    // Подпись поля в строке-сводке свёрнутой секции («Vн», «H₀»); без неё
    // сводка берёт `name`.
    shortLabel: z.string().min(1).max(40).optional(),
    hint: z.string().min(1).max(160).optional(),
    // Пояснение к варианту ряда («1» → «h 1.5; t123 13.5»); в submit уходит
    // само значение, не текст.
    optionNotes: z.record(z.string(), z.string()).optional(),
    // Подсказки, зависящие от значения другого поля (ряды Прил. Е разные для
    // жилых и офисов). Значение, выпавшее из нового ряда, не сбрасывается.
    optionsBy: z
      .object({
        source: z.string().min(1).max(80),
        scope: liftEditorFieldScopeSchema.optional(),
        options: z.record(z.string(), z.array(z.string()).min(1)),
      })
      .strict()
      .optional(),
  })
  .strict();

/**
 * Декларативное зависимое значение: при изменении любого источника ищется
 * строка, чей `when` совпал со значениями `sources` по порядку, и её `set`
 * перезаписывает поля лифта. Нормативных таблиц ГОСТ в компоненте нет —
 * приходят готовые строки (Решение 7 design.md).
 */
export const liftEditorDependentRuleSchema = z
  .object({
    sources: z
      .array(
        z
          .object({
            field: z.string().min(1).max(80),
            scope: liftEditorFieldScopeSchema.optional(),
          })
          .strict(),
      )
      .min(1)
      .max(4),
    rows: z
      .array(
        z
          .object({
            // Порядок значений = порядок `sources`.
            when: z.array(z.union([z.string(), z.number()])).min(1),
            set: z.record(z.string(), z.union([z.string(), z.number()])),
          })
          .strict(),
      )
      .min(1),
    // Нет совпадения: 'keep' — оставить как есть (свой Vn не трогает h/t123),
    // 'clear' — очистить и дать ввести вручную. Default 'keep'.
    onNoMatch: z.enum(['keep', 'clear']).optional(),
  })
  .strict();

/** Конфиг одной методики: состав экранов, режим лифтов и правила подстановок. */
export const liftEditorMethodConfigSchema = z
  .object({
    // Ключ методики в домене агента ('52941' | '34758'); компонент только сверяет строки.
    method: z.string().min(1).max(40),
    label: z.string().min(1).max(120),
    // Подзаголовок экрана; логики за ним нет.
    gostLabel: z.string().min(1).max(60),
    buildingTitle: z.string().min(1).max(120),
    buildingFields: z.array(liftEditorFieldSchema).min(1),
    liftTitle: z.string().min(1).max(120),
    liftFields: z.array(liftEditorFieldSchema).min(1),
    liftsMode: liftsModeSchema,
    // 'Лифт' → «Лифт 1»; в режиме группы — подпись единственной секции.
    liftTabLabel: z.string().min(1).max(80),
    // Тип здания для текста шапки «gostLabel · тип», когда у ветки нет поля
    // `buildingType` («жилое здание» для 52941).
    buildingKindLabel: z.string().min(1).max(80).optional(),
    maxLifts: z.number().int().min(1).max(16).optional(),
    dependentRules: z.array(liftEditorDependentRuleSchema).max(8).optional(),
  })
  .strict();

export const liftEditorPropsSchema = z
  .object({
    // Управляет активным конфигом, поэтому отдельный проп, а не первое поле
    // `buildingFields` (Решение 3 design.md). Рендерится первым на «Здании».
    methodField: liftEditorFieldSchema,
    // Конфиги ВСЕХ методик сразу — смена методики локальна, без раунд-трипа.
    methodConfigs: z.array(liftEditorMethodConfigSchema).min(1),
    method: z.string().min(1).max(40),
    // Значения активной ветки; черновики остальных живут только на клиенте.
    building: z.record(z.string(), z.unknown()),
    lifts: z.array(z.record(z.string(), z.unknown())),
    buildingTabLabel: z.string().min(1).max(80),
    advancedLabel: z.string().min(1).max(80),
    addLabel: z.string().min(1).max(80),
    removeLabel: z.string().min(1).max(80),
    submitLabel: z.string().min(1).max(80),
    // Единственный action компонента (Решение 10 design.md).
    submitAction: z.string().min(1).max(120),
    // Имя action'а автосохранения черновика. Необязательный: без него компонент
    // ведёт себя как раньше — наружу уходит только submit.
    draftAction: z.string().min(1).max(120).optional(),
    // Шапка карточки: заголовок слева, контекст и переключатель методики
    // справа. Без `headerTitle` шапки нет — переключатель падает над секции.
    headerTitle: z.string().min(1).max(120).optional(),
    headerContext: z.string().min(1).max(200).optional(),
    // Подпись кнопки-навигации «Далее»: пока есть незаполненные обязательные
    // поля или непросмотренные секции, кнопка ведёт по секциям и не шлёт
    // действий. Без пропа кнопка блокируется, как раньше.
    pendingLabel: z.string().min(1).max(80).optional(),
    // Источники значений полей — только подписи под контролами; ничего не
    // блокируют и в payload не уходят. `liftSources` — по индексам лифтов.
    buildingSources: liftEditorSectionSourcesSchema.optional(),
    liftSources: z.array(liftEditorSectionSourcesSchema).optional(),
  })
  .strict()
  .superRefine((value, ctx) => {
    if (!value.methodConfigs.some(config => config.method === value.method)) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: `methodConfigs has no config for method "${value.method}"`,
        path: ['method'],
      });
    }

    if (value.methodField.type !== 'select') {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: 'methodField.type must be "select"',
        path: ['methodField', 'type'],
      });
    }

    for (const [configIndex, config] of value.methodConfigs.entries()) {
      for (const [ruleIndex, rule] of (config.dependentRules ?? []).entries()) {
        for (const [rowIndex, row] of rule.rows.entries()) {
          if (row.when.length !== rule.sources.length) {
            ctx.addIssue({
              code: z.ZodIssueCode.custom,
              message: `Row "when" has ${row.when.length} value(s) but the rule declares ${rule.sources.length} source(s).`,
              path: [
                'methodConfigs',
                configIndex,
                'dependentRules',
                ruleIndex,
                'rows',
                rowIndex,
                'when',
              ],
            });
          }
        }
      }
    }
  });

export type LiftsMode = z.infer<typeof liftsModeSchema>;
export type LiftEditorFieldSource = z.infer<typeof liftEditorFieldSourceSchema>;
export type LiftEditorSectionSources = z.infer<typeof liftEditorSectionSourcesSchema>;
export type LiftEditorFieldScope = z.infer<typeof liftEditorFieldScopeSchema>;
export type LiftEditorField = z.infer<typeof liftEditorFieldSchema>;
export type LiftEditorDependentRule = z.infer<typeof liftEditorDependentRuleSchema>;
export type LiftEditorMethodConfig = z.infer<typeof liftEditorMethodConfigSchema>;
export type LiftEditorProps = z.infer<typeof liftEditorPropsSchema>;

export const liftEditorDefinition: CatalogComponentDefinition<typeof liftEditorPropsSchema> = {
  name: 'LiftEditor',
  slug: 'lift-editor',
  description:
    'A single-screen editor for a lift (elevator) calculation document: a card header with the calculation-method switcher ("GOST · building kind"), a collapsible building section on top and lift sections below it (or a single lift-group section, depending on the method). Collapsed sections show a live summary line built from current values; a two-mode footer button navigates to the next incomplete/unreviewed section ("Next") and turns into submit once the document has been walked through. Everything happens on the client — every method config, option row and derived-value table is preloaded in props — and the whole document ({method, building, lifts}) goes back to the agent in a single action. Optional provenance blocks (buildingSources/liftSources) caption fields with where their values came from; edits are auto-saved as debounced drafts when draftAction is set. Fields support free text entry with normative-row hints (combo), option lists driven by another field, declarative dependent values, and a collapsible summarised group for defaulted parameters.',
  schema: liftEditorPropsSchema,
};

/**
 * Тот же экран подбора лифтов на новом наборе примитивов
 * (`packages/catalog-react/src/primitives`, интерактивность — `@base-ui/react`).
 * Схема props общая с `liftEditorDefinition` намеренно: сравнивать «было /
 * стало» на разном наполнении бессмысленно, поэтому новое имя нужно только
 * компоненту, не данным. Оба рендерера зарегистрированы одновременно — агент
 * может адресовать любой.
 */
export const liftEditorNextDefinition: CatalogComponentDefinition<typeof liftEditorPropsSchema> = {
  name: 'LiftEditorNext',
  slug: 'lift-editor-next',
  description:
    'The same lift (elevator) calculation editor as `LiftEditor` — identical props, identical data contract, both calculation methods with the same client-side form rebuild — rendered on the catalog primitive set (cards, buttons, chips, one type scale, tokenised colours) with interaction taken from a headless component library: accordion sections that open with Enter/Space and expose aria-expanded/aria-controls, dropdowns, number fields with arrow-key stepping, and free-text fields with a normative-row hint list. Prefer it when the surface should be navigable without a mouse; emit the same props as for `LiftEditor`.',
  schema: liftEditorPropsSchema,
};

File: packages/catalog-python/src/ai37_a2ui_catalog/models/lift_editor.py
from __future__ import annotations

from typing import Annotated, Any, Literal

from pydantic import Field, model_validator

from .form_card import FormFieldLookupValue, LookupSuggestMode
from .shared import StrictModel

# Режим лифтов методики: перечень вкладок «Лифт 1…N» либо одна лифтовая группа.
LiftsMode = Literal["per-lift", "group"]

# Экран поля-источника: значения лифта (по умолчанию) или здания.
LiftEditorFieldScope = Literal["building", "lift"]

# `combo` — свободный ввод с подсказками ряда; остальное — типы поля FormCard.
LiftEditorFieldType = Literal["text", "number", "select", "boolean", "lookup", "combo"]

# Источник значения поля — тот же контракт, что `generalSources` теплотеха.
LiftEditorFieldSourceKind = Literal["project", "question", "suggested", "default"]


class LiftEditorFieldSource(StrictModel):
    source: LiftEditorFieldSourceKind
    # Человеческое обоснование одной строкой.
    note: str = Field(default=None, min_length=1, max_length=200)


class LiftEditorFieldOptionsBy(StrictModel):
    """Подсказки, зависящие от значения другого поля (ряды Прил. Е по типу здания)."""

    source: str = Field(min_length=1, max_length=80)
    scope: LiftEditorFieldScope = None
    options: dict[str, Annotated[list[str], Field(min_length=1)]]


class LiftEditorField(StrictModel):
    """Поле редактора: базовое поле FormCard плюс аддитивные ключи."""

    name: str = Field(min_length=1, max_length=80)
    label: str = Field(min_length=1, max_length=120)
    type: LiftEditorFieldType
    required: bool = None
    options: list[str] = Field(default=None, min_length=1)
    referenceId: str = Field(default=None, min_length=1, max_length=80)
    minChars: int = Field(default=None, ge=1, le=10)
    suggestMode: LookupSuggestMode = None
    placeholder: str = Field(default=None, min_length=1, max_length=120)
    defaultValue: str | int | float | bool | FormFieldLookupValue = None
    # Поле со значением по умолчанию — под экспандер.
    advanced: bool = None
    # Подпись поля в строке-сводке свёрнутой секции; без неё сводка берёт `name`.
    shortLabel: str = Field(default=None, min_length=1, max_length=40)
    hint: str = Field(default=None, min_length=1, max_length=160)
    # Пояснение к варианту ряда; в submit уходит само значение.
    optionNotes: dict[str, str] = None
    optionsBy: LiftEditorFieldOptionsBy = None


class LiftEditorDependentRuleSource(StrictModel):
    field: str = Field(min_length=1, max_length=80)
    scope: LiftEditorFieldScope = None


class LiftEditorDependentRuleRow(StrictModel):
    # Порядок значений = порядок `sources` правила.
    when: list[str | float] = Field(min_length=1)
    set: dict[str, str | float]


class LiftEditorDependentRule(StrictModel):
    """Декларативное зависимое значение: готовые строки ряда, без таблиц ГОСТ."""

    sources: list[LiftEditorDependentRuleSource] = Field(min_length=1, max_length=4)
    rows: list[LiftEditorDependentRuleRow] = Field(min_length=1)
    # Нет совпадения: 'keep' (default) оставляет как есть, 'clear' очищает.
    onNoMatch: Literal["keep", "clear"] = None


class LiftEditorMethodConfig(StrictModel):
    """Конфиг одной методики: состав экранов, режим лифтов и правила подстановок."""

    method: str = Field(min_length=1, max_length=40)
    label: str = Field(min_length=1, max_length=120)
    gostLabel: str = Field(min_length=1, max_length=60)
    buildingTitle: str = Field(min_length=1, max_length=120)
    buildingFields: list[LiftEditorField] = Field(min_length=1)
    liftTitle: str = Field(min_length=1, max_length=120)
    liftFields: list[LiftEditorField] = Field(min_length=1)
    liftsMode: LiftsMode
    liftTabLabel: str = Field(min_length=1, max_length=80)
    # Тип здания для текста шапки, когда у ветки нет поля `buildingType`.
    buildingKindLabel: str = Field(default=None, min_length=1, max_length=80)
    maxLifts: int = Field(default=None, ge=1, le=16)
    dependentRules: list[LiftEditorDependentRule] = Field(default=None, max_length=8)


class LiftEditorProps(StrictModel):
    # Управляет активным конфигом, поэтому отдельный проп, а не поле здания.
    methodField: LiftEditorField
    # Конфиги ВСЕХ методик сразу — смена методики локальна, без раунд-трипа.
    methodConfigs: list[LiftEditorMethodConfig] = Field(min_length=1)
    method: str = Field(min_length=1, max_length=40)
    # Значения активной ветки; черновики остальных живут только на клиенте.
    building: dict[str, Any]
    lifts: list[dict[str, Any]]
    buildingTabLabel: str = Field(min_length=1, max_length=80)
    advancedLabel: str = Field(min_length=1, max_length=80)
    addLabel: str = Field(min_length=1, max_length=80)
    removeLabel: str = Field(min_length=1, max_length=80)
    submitLabel: str = Field(min_length=1, max_length=80)
    # Единственный action компонента.
    submitAction: str = Field(min_length=1, max_length=120)
    # Имя action'а автосохранения черновика; без него автосейва нет.
    draftAction: str = Field(default=None, min_length=1, max_length=120)
    # Шапка карточки; без `headerTitle` переключатель методики падает над секции.
    headerTitle: str = Field(default=None, min_length=1, max_length=120)
    headerContext: str = Field(default=None, min_length=1, max_length=200)
    # Подпись кнопки-навигации «Далее»; без пропа кнопка блокируется, как раньше.
    pendingLabel: str = Field(default=None, min_length=1, max_length=80)
    # Источники значений полей — только подписи; в payload не уходят.
    buildingSources: dict[str, LiftEditorFieldSource] = None
    liftSources: list[dict[str, LiftEditorFieldSource]] = None

    @model_validator(mode="after")
    def validate_document(self) -> "LiftEditorProps":
        if all(config.method != self.method for config in self.methodConfigs):
            raise ValueError(f'methodConfigs has no config for method "{self.method}"')

        if self.methodField.type != "select":
            raise ValueError('methodField.type must be "select"')

        for config in self.methodConfigs:
            for rule in config.dependentRules or []:
                for row in rule.rows:
                    if len(row.when) != len(rule.sources):
                        raise ValueError(
                            f'Row "when" has {len(row.when)} value(s) but the rule '
                            f"declares {len(rule.sources)} source(s)."
                        )

        return self
